matrix-mult.futil

```diff
@@ -0,0 +1,234 @@
+      mul_upd[done] = temp1.done;
+    }
+
+    group mul_upd2 { //multiply 01 x 10, store in temp2
+      mul.left = val3.out;
+      mul.right = val6.out;
+      mul.go = !mul.done ? 1'd1;
+      temp2.in = mul.out;
+      temp2.write_en = mul.done;
+      mul_upd2[done] = temp2.done;
+    }
+
+    group upd { // add together and store in val
+      add.left = temp1.out;
+      add.right = temp2.out;
+      val.in = add.out;
+      val.write_en = 1'b1;
+      upd[done] = val.done;
+    }
+
+    group mul_upd3 { // multiply 00 x 01, store in temp1
+      mul.left = val1.out;
+      mul.right = val4.out;
+      mul.go = !mul.done ? 1'd1;
+      temp1.in = mul.out;
+      temp1.write_en = mul.done;
+      mul_upd3[done] = temp1.done;
+    }
+
+    group mul_upd4 { // multiply 01 x 11, store in temp2
+      mul.left = val3.out;
+      mul.right = val8.out;
+      mul.go = !mul.done ? 1'd1;
+      temp2.in = mul.out;
+      temp2.write_en = mul.done;
+      mul_upd4[done] = temp2.done;
+    }
+
+    group mul_upd5 { // multiply 10 x 00, store in temp1
+      mul.left = val5.out;
+      mul.right = val2.out;
+      mul.go = !mul.done ? 1'd1;
+      temp1.in = mul.out;
+      temp1.write_en = mul.done;
+      mul_upd5[done] = temp1.done;
+    }
+
+    group mul_upd6 { // multiply 11 x 10, store in temp2
+      mul.left = val7.out;
+      mul.right = val6.out;
+      mul.go = !mul.done ? 1'd1;
+      temp2.in = mul.out;
+      temp2.write_en = mul.done;
+      mul_upd6[done] = temp2.done;
+    }
+
+    group mul_upd7 { // multiply 10 x 01, store in temp1
+      mul.left = val5.out;
+      mul.right = val4.out;
+      mul.go = !mul.done ? 1'd1;
+      temp1.in = mul.out;
+      temp1.write_en = mul.done;
+      mul_upd7[done] = temp1.done;
+    }
+
+    group mul_upd8 { // multiply 11 x 11, store in temp2
+      mul.left = val7.out;
+      mul.right = val8.out;
+      mul.go = !mul.done ? 1'd1;
+      temp2.in = mul.out;
+      temp2.write_en = mul.done;
+      mul_upd8[done] = temp2.done;
+    }
+
+    group write { //write to position 00 in result matrix, do at very end
+      mem1.addr0 = 1'b0;
+      mem1.addr1 = 1'b0;
+      mem1.write_en = 1'b1;
+      mem1.write_data = val.out;
+      write[done] = mem1.done;
+    }
+
+    group write1 { //write to position 10 in result matrix, do at very end
+      mem1.addr0 = 1'b1;
+      mem1.addr1 = 1'b0;
+      mem1.write_en = 1'b1;
+      mem1.write_data = val.out;
+      write1[done] = mem1.done;
+    }
+    group write2 { //write to position 01 in result matrix, do at very end
+      mem1.addr0 = 1'b0;
+      mem1.addr1 = 1'b1;
+      mem1.write_en = 1'b1;
+      mem1.write_data = val.out;
+      write2[done] = mem1.done;
+    }
+
+    group write3 { //write to position 11 in result matrix, do at very end
+      mem1.addr0 = 1'b1;
+      mem1.addr1 = 1'b1;
+      mem1.write_en = 1'b1;
+      mem1.write_data = val.out;
+      write3[done] = mem1.done;
+    }
+
+  }
+  control {
+    seq {
+      read1;
```

Comment:
You may have some "fun" doing the mental gymnastics to figure out which pieces of your program can actually be done in parallel! If you try to read `mem1[0][0]` and some other cell of `mem1` (e.g. `mem1[0][1]` at the same time, that's a bad idea... `mem1` won't know if its `addr1` port is meant to be `0` or `1`. But I bet you can read `mem1[0][0]` at the same time as `mem2[0][0]`. There are other opportunities too. We can look at those together, and we'll also look at a Calyx pass that can find these opportunities for you automatically!

Reply:
Just putting this here for easy reference later:
```calyx matrix.futil -p cell-share -m file > matrix_shared.futil```
will run the cell-sharing pass for you

Reply:
And I've begun to ask what pass exploits opportunities for parallelism

Reply:
`static-promotion`, apparently. Hmm, it may be a bit of a noisy comparison but we'll try!